### PR TITLE
Add support for NEAR Protocol (NEAR) addresses

### PR DESCRIPTION
**Pull request by janus**
## Issue number

Closes https://github.com/ensdomains/address-encoder/issues/215

## Description
Add two function to handle bs58Encode and bs58Decode. These are no checked.

## List of features added/changed
 Just specify bs58EncodeNoCheck and  bs58DecodeNoCheck for NEAR in configuration
Reference 
[1] https://github.com/near/near-api-js/blob/79a336931943384708a644d9ca4ce327a1daec07/src/utils/key_pair.ts#L34
[2] https://github.com/near/near-api-js/blob/master/src/utils/serialize.ts#L9

## How Has This Been Tested?
Add NEAR in the test pool so that decoded address should match it text version. 

## Checklist:

- [X ] My code follows the code style of this project.
- [ X] My code implements all the required features.

### CODE REVIEW

**Review by makoto — CHANGES_REQUESTED** [review]
Please provide a reference link to the specification

**Review by janus — COMMENTED** [review]
Resolved

**Comment by makoto** [comment]
> Resolved

What did you resolve? It's still missing the reference link to the spec

**Comment by janus** [comment]
Thanks. Resolved.

**Comment by ilblackdragon** [comment]
If the question about addresses / account ids in NEAR - they are NOT base58.
Here is full spec for account ids - https://nomicon.io/DataStructures/Account.html#account-id

Here is a regex: ^(([a-z\d]+[\-_])*[a-z\d]+\.)*([a-z\d]+[\-_])*[a-z\d]+$
+ restriction of [2; 64] characters.

**Comment by makoto** [comment]
Hi, please let me know if you will continue working on this or not. If not we will close this PR

**Comment by janus** [comment]
> Hi, please let me know if you will continue working on this or not. If not we will close this PR

I would come back to this next month

**Comment by brantlymillegan** [comment]
@janus You still interested in doing this? Would love to get this added sooner rather than later

**Comment by janus** [comment]
@245      Looks like you were quick in closing out this. I still want to complete it

**Comment by brantlymillegan** [comment]
> @245 Looks like you were quick in closing out this. I still want to complete it

Reopened

**Comment by janus** [comment]
> If the question about addresses / account ids in NEAR - they are NOT base58.
> Here is full spec for account ids - https://nomicon.io/DataStructures/Account.html#account-id
> 
> Here is a regex: `^(([a-z\d]+[\-_])*[a-z\d]+\.)*([a-z\d]+[\-_])*[a-z\d]+$` + restriction of [2; 64] characters.

I think what I did is correct. Please read down the link you posted.


Implicit account IDs

Implicit accounts work similarly to Bitcoin/Ethereum accounts. It allows you to reserve an account ID before it's created by generating a ED25519 key-pair locally. This key-pair has a public key that maps to the account ID. The account ID is a lowercase hex representation of the public key. ED25519 Public key is 32 bytes that maps to 64 characters account ID.

Example: public key in base58 BGCCDDHfysuuVnaNVtEhhqeT4k9Muyem3Kpgq2U1m9HX will map to an account ID 98793cd91a3f870fb126f66285808c7e094afcfc4eda8a970f6648cdf0dbd6de.

The corresponding secret key allows you to sign transactions on behalf of this account once it's created on chain.
Implicit account creation

An account with implicit account ID can only be created by sending a transaction/receipt with a single Transfer action to the implicit account ID receiver:

    The account will be created with the account ID.
    The account will have a new full access key with the ED25519-curve public key of decode_hex(account_id) and nonce 0.
    The account balance will have a transfer balance deposited to it.

This account can not be created using CreateAccount action to avoid being able to hijack the account without having the corresponding private key.

**Comment by janus** [comment]
@brantlymillegan  Please review and comment.  Example from https://nomicon.io/DataStructures/Account.html

**Comment by ilblackdragon** [comment]
@janus You are correct about implicit accounts being representations of the keys.

But Implicit accounts are a **subset** of all accounts.
Why would you limit what types of accounts can be used here?

The general regex matches all accounts in NEAR.
Also even if you only want to represent implicit accounts only - they are hex encoded 64 characters.

So I'm not sure why would public keys / base58 would be used here?

**Comment by makoto** [comment]
Hi, @janus I suggest you follow the implementation of TrustWallet so that it's compatible across other wallets.

https://github.com/trustwallet/wallet-core/blob/master/src/NEAR/Address.cpp
https://github.com/trustwallet/wallet-core/blob/master/src/NEAR/Account.cpp
https://github.com/trustwallet/wallet-core/blob/master/tests/NEAR/AddressTests.cpp
https://github.com/trustwallet/wallet-core/blob/master/tests/NEAR/AccountTests.cpp

**Comment by janus** [comment]
> Hi, @janus I suggest you follow the implementation of TrustWallet so that it's compatible across other wallets.
> 
> https://github.com/trustwallet/wallet-core/blob/master/src/NEAR/Address.cpp
> https://github.com/trustwallet/wallet-core/blob/master/src/NEAR/Account.cpp
> https://github.com/trustwallet/wallet-core/blob/master/tests/NEAR/AddressTests.cpp
> https://github.com/trustwallet/wallet-core/blob/master/tests/NEAR/AccountTests.cpp

Thanks. Noted.

**Comment by janus** [comment]
> @janus You are correct about implicit accounts being representations of the keys.
> 
> But Implicit accounts are a **subset** of all accounts.
> Why would you limit what types of accounts can be used here?
> 
> The general regex matches all accounts in NEAR.
> Also even if you only want to represent implicit accounts only - they are hex encoded 64 characters.
> 
> So I'm not sure why would public keys / base58 would be used here?
> 
> Like looking at other blockchains represented here, they are all encoding the address not the public key (e.g. 20 hex character address for ETH, not 32 character public key).

Thanks. Noted.

**Comment by janus** [comment]
> Hi, @janus I suggest you follow the implementation of TrustWallet so that it's compatible across other wallets.
> 
> https://github.com/trustwallet/wallet-core/blob/master/src/NEAR/Address.cpp
> https://github.com/trustwallet/wallet-core/blob/master/src/NEAR/Account.cpp
> https://github.com/trustwallet/wallet-core/blob/master/tests/NEAR/AddressTests.cpp
> https://github.com/trustwallet/wallet-core/blob/master/tests/NEAR/AccountTests.cpp

I noticed that in trustwallet NEAR Account and NEAR Address are not same thing. Address  is a valid account that may start with 'NEAR' . Bs58 is used to decode address and 4 last bytes removed. Another form is just the product of the first form and it is checked for its size/length, which is based on PublicKey::ed25519Size(32).

trustwallet just matched NEAR regex with Account ID and limit size within 2 and 64. 

I chose trustwallet Address format.  Please I need your input here.

**Comment by janus** [comment]
@ilblackdragon  Please review and comment.

**Comment by hewigovens** [comment]
No need to support base58 address starts with “NEAR”, that’s deprecated; you only need to support implicit account (hex string) and xxx.near

**Comment by janus** [comment]
> No need to support base58 address starts with “NEAR”, that’s deprecated; you only need to support implicit account (hex string) and xxx.near

@hewigovens  @makoto  I just followed the above instruction. Please review and merge.

File: src/index.ts [hunk 1]
```diff
@@ -1061,6 +1098,7 @@ export const formats: IFormat[] = [
   eosioChain('FIO', 235, 'FIO'),
   getConfig('BSV', 236, bsvAddresEncoder, bsvAddressDecoder),
   getConfig('NEO', 239, bs58Encode, bs58Decode),
+  getConfig('NEAR', 242, encodeNearAddr, decodeNearAddr),
```
Comment (makoto) [inline src/index.ts:1101]: slip44 id is incorrect

File: src/__tests__/index.test.ts [hunk 2]
```diff
@@ -438,6 +438,16 @@ const vectors: Array<TestVector> = [
     coinType: 239,
     passingVectors: [{ text: 'AXaXZjZGA3qhQRTCsyG5uFKr9HeShgVhTF', hex: '17ad5cac596a1ef6c18ac1746dfd304f93964354b5' }],
   },
+  {
+    name: 'NEAR',
+    coinType: 242,
```
Comment (makoto) [inline src/__tests__/index.test.ts:443]: slip44 id is incorrect

**Review by makoto — CHANGES_REQUESTED** [review]
Please add correct slip44 id

File: src/__tests__/index.test.ts (hunk 2 above)
Comment (janus) [reply src/__tests__/index.test.ts]: Thanks. Resolved. Please review and merge.

File: src/index.ts (hunk 1 above)
Comment (janus) [reply src/index.ts]: @makoto  Done. Review. Merge.

File: README.md [hunk 3]
```diff
@@ -93,6 +93,7 @@ This library currently supports the following cryptocurrencies and address forma
  - LUNA (bech32)
  - MONA (base58check P2PKH and P2SH, and bech32 segwit)
  - NANO (nano-base32)
+ - NEAR (base58, no check)
```
Comment (ilblackdragon) [inline README.md:96]: plain text? or just remove this comment?

File: src/index.ts [hunk 4]
```diff
@@ -229,6 +229,43 @@ function encodeCashAddr(data: Buffer): string {
...
+    throw Error('Invalid address string');
+  }
+
+  let ndata  = data;
+  if(data.startsWith('0x')) {
+    ndata = data.substr(2);
+  }
+  
+  if(data.length >= 2 || data.length <= 64) {  
+    if((data.endsWith('.near') && data.length > 5)) {
```
Comment (ilblackdragon) [inline src/index.ts:245]: hm, the account id "blah.com" is also valid.
Top Level Accounts are coming at some point in 2021 -- https://nomicon.io/DataStructures/Account.html#top-level-accounts

File: src/__tests__/index.test.ts [hunk 5]
```diff
@@ -549,6 +549,16 @@ const vectors: Array<TestVector> = [
       { text: 'ABBC5MTKdW6dFqEjYqQYMmLohCsALWcBAx2xRapzDTKAtz3XwKJcaf', hex: '023d3a2e33a90f8f5bcbda1ec129ba1eee5e5f2ab6a77d652cbb0517f2b49669e8' },
     ],
   },
+  {
+    name: 'NEAR',
+    coinType: 397,
+    passingVectors: [
+      { text: '9902c136629fc630416e50d4f2fef6aff867ea7e.lockup.near',
```
Comment (ilblackdragon) [inline src/__tests__/index.test.ts:556]: I would add something like "test.evm" just to check it's not only `.near` TLA

File: src/index.ts (hunk 4 above)
Comment (janus) [reply src/index.ts]: It can accept ".com".

File: README.md (hunk 3 above)
Comment (janus) [reply README.md]: Resolved. Thanks

**Comment by janus** [comment]
@makoto  @hewigovens  @ilblackdragon  Let's have a common agreement on what we want and stick to it. I would like to complete this PR this week. I have budgeted the money from it already for Xmas celebration :)

File: src/index.ts (hunk 4 above)
Comment (makoto) [reply src/index.ts]: I think @ilblackdragon meant that. `.com` is also acceptable name, not just `.near` and `.com`. In their doc,  `over.9000` and `a.bro` are also added as "examples" of acceptable domains. I don't think you need suffix check at all.

File: src/index.ts [hunk 6]
```diff
@@ -229,6 +229,43 @@ function encodeCashAddr(data: Buffer): string {
...
+  } else {
+    throw Error('Invalid address string');
+  }  
+}
+
+function encodeNearAddr(data: Buffer): string {
+  const ndata = data.toString();
+  if(ndata.length > 64 || ndata.length < 2) {
+    throw Error('Invalid address format');
+
```
Comment (makoto) [inline src/index.ts:264]: Please remove extra line

**Comment by makoto** [comment]
@janus according to the https://nomicon.io/DataStructures/Account.html#top-level-accounts, the rule is quite clear.

```
Account ID Rules
minimum length is 2
maximum length is 64
Account ID consists of Account ID parts separated by .
Account ID part consists of lowercase alphanumeric symbols separated by either _ or -.
Account ID that is 64 characters long consists of lowercase hex characters is a specific implicit account ID.
```

The internal account is the subset of this rule so you do't need custom logic for handling it, and the rule does not state that you need suffix check either.

File: src/index.ts (hunk 6 above)
Comment (janus) [reply src/index.ts]: Resolved

**Comment by janus** [comment]
> @janus according to the https://nomicon.io/DataStructures/Account.html#top-level-accounts, the rule is quite clear.
> 
> ```
> Account ID Rules
> minimum length is 2
> maximum length is 64
> Account ID consists of Account ID parts separated by .
> Account ID part consists of lowercase alphanumeric symbols separated by either _ or -.
> Account ID that is 64 characters long consists of lowercase hex characters is a specific implicit account ID.
> ```
> 
> The internal account is the subset of this rule so you do't need custom logic for handling it, and the rule does not state that you need suffix check either.

Done. Review.

**Pull request merged**